File: typescript/packages/sdk/src/http-client.ts
/**
 * HTTP client for hosted mode communication with the ECP server.
 *
 * Uses native fetch (Node 18+) with retry logic and HMAC-SHA256 signing.
 */

import crypto from 'node:crypto';
import type { Decision } from '@hiitl/core';
import type { HostedModeConfig } from './config.js';
import { ServerError, NetworkError } from './exceptions.js';

/** Status codes that trigger a retry. */
const RETRYABLE_STATUS_CODES = new Set([429, 502, 503, 504]);

/**
 * Hosted mode HTTP client for communicating with the ECP server.
 */
export class HostedClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly timeout: number;
  private readonly maxRetries: number;
  private readonly signatureKey?: string;
  private readonly agentId: string;
  private readonly orgId: string;
  private readonly environment: string;

  constructor(config: HostedModeConfig) {
    this.baseUrl = config.server_url;
    this.apiKey = config.api_key;
    this.timeout = config.timeout;
    this.maxRetries = config.max_retries;
    this.signatureKey = config.signature_key;
    this.agentId = config.agent_id;
    this.orgId = config.org_id;
    this.environment = config.environment;
  }

  /**
   * Send an evaluate request to the ECP server.
   *
   * @param options - Evaluation parameters
   * @returns Decision object from the server
   * @throws {ServerError} On non-2xx responses
   * @throws {NetworkError} On connection failures
   */
  async evaluate(options: {
    action: string;
    operation: string;
    target: Record<string, unknown>;
    parameters: Record<string, unknown>;
    agent_id?: string;
    user_id?: string;
    session_id?: string;
    confidence?: number;
    reason?: string;
    sensitivity?: string[];
    cost_estimate?: { tokens?: number; usd_cents?: number };
    idempotency_key?: string;
  }): Promise<Decision> {
    // Build request body
    const body: Record<string, unknown> = {
      action: options.action,
      operation: options.operation,
      target: options.target,
      parameters: options.parameters,
      agent_id: options.agent_id ?? this.agentId,
    };

    // Add optional fields only if provided
    if (options.user_id !== undefined) body.user_id = options.user_id;
    if (options.session_id !== undefined) body.session_id = options.session_id;
    if (options.confidence !== undefined) body.confidence = options.confidence;
    if (options.reason !== undefined) body.reason = options.reason;
    if (options.sensitivity !== undefined) body.sensitivity = options.sensitivity;
    if (options.cost_estimate !== undefined) body.cost_estimate = options.cost_estimate;
    if (options.idempotency_key !== undefined) body.idempotency_key = options.idempotency_key;

    // Compute signature if key is configured
    if (this.signatureKey) {
      body.signature = this._computeSignature(body);
    }

    // Send with retry
    const response = await this._sendWithRetry(body);
    return this._parseResponse(response);
  }

  /**
   * Compute HMAC-SHA256 signature of the request body.
   */
  private _computeSignature(body: Record<string, unknown>): string {
    const canonical = JSON.stringify(body, Object.keys(body).sort());
    return crypto
      .createHmac('sha256', this.signatureKey!)
      .update(canonical)
      .digest('hex');
  }

  /**
   * Send request with exponential backoff retry on transient failures.
   */
  private async _sendWithRetry(body: Record<string, unknown>): Promise<Response> {
    const url = `${this.baseUrl}/v1/evaluate`;
    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = backoffDelay(attempt - 1);
        await sleep(delay);
      }

      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${this.apiKey}`,
          },
          body: JSON.stringify(body),
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        // Don't retry on non-retryable status codes
        if (!RETRYABLE_STATUS_CODES.has(response.status)) {
          return response;
        }

        // Retryable status — continue loop if retries remain
        lastError = new Error(`Server returned ${response.status}`);
      } catch (err: any) {
        // AbortError = timeout, TypeError = network failure — both retryable
        if (err.name === 'AbortError') {
          lastError = new Error(`Request timed out after ${this.timeout}ms`);
        } else {
          lastError = err;
        }

        // Non-retryable fetch errors (not timeout, not network)
        if (err.name !== 'AbortError' && err.name !== 'TypeError') {
          throw new NetworkError(this.baseUrl, err);
        }
      }
    }

    // All retries exhausted
    throw new NetworkError(this.baseUrl, lastError ?? new Error('Request failed'));
  }

  /**
   * Parse server response into a Decision object.
   */
  private async _parseResponse(response: Response): Promise<Decision> {
    if (response.status === 200) {
      const data = (await response.json()) as Record<string, any>;

      // Bridge server timing format to SDK Timing format
      const rawTiming = data.timing ?? {};
      const totalMs = rawTiming.total_ms ?? 0;

      const decision: Decision = {
        action_id: data.action_id ?? '',
        decision: data.decision,
        allowed: data.allowed,
        reason_codes: data.reason_codes ?? [],
        policy_version: data.policy_version ?? '',
        timing: {
          ingest_ms: rawTiming.ingest_ms ?? 0,
          evaluation_ms: rawTiming.evaluation_ms ?? totalMs,
          total_ms: totalMs,
        },
      };

      // Add optional fields if present
      if (data.envelope_hash !== undefined && data.envelope_hash !== null) {
        (decision as any).envelope_hash = data.envelope_hash;
      }
      if (data.resume_token) decision.resume_token = data.resume_token;
      if (data.route_ref) decision.route_ref = data.route_ref;
      if (data.escalation_context) decision.escalation_context = data.escalation_context;
      if (data.matched_rules) decision.matched_rules = data.matched_rules;
      if (data.rate_limit) decision.rate_limit = data.rate_limit;
      if (data.approval_metadata) decision.approval_metadata = data.approval_metadata;
      if (data.sandbox_metadata) decision.sandbox_metadata = data.sandbox_metadata;
      if (data.error) decision.error = data.error;
      if (data.remediation) decision.remediation = data.remediation;
      if (data.would_be) decision.would_be = data.would_be;
      if (data.would_be_reason_codes) decision.would_be_reason_codes = data.would_be_reason_codes;

      return decision;
    }

    // Error responses
    let errorCode = 'UNKNOWN_ERROR';
    let errorMessage = `Server returned status ${response.status}`;

    try {
      const data = (await response.json()) as Record<string, any>;
      if (typeof data.detail === 'string') {
        errorCode = data.detail;
        errorMessage = data.detail;
      } else if (data.detail?.code) {
        errorCode = data.detail.code;
        errorMessage = data.detail.message ?? data.detail.code;
      } else if (data.error) {
        errorCode = data.error;
        errorMessage = data.message ?? data.error;
      }
    } catch {
      // Could not parse response body — use defaults
    }

    throw new ServerError(response.status, errorCode, errorMessage);
  }
}

/**
 * Compute exponential backoff delay: min(0.5 * 2^attempt, 4) seconds.
 * Returns delay in milliseconds.
 */
export function backoffDelay(attempt: number): number {
  return Math.min(500 * Math.pow(2, attempt), 4000);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: python/hiitl/sdk/http_client.py
"""HTTP client for hosted mode ECP server communication.

Handles:
- POST /v1/evaluate with Bearer token auth
- Retry with exponential backoff on transient failures (5xx, timeouts)
- Response parsing to Decision objects
- Error mapping to SDK exceptions with helpful messages
"""

import hashlib
import hmac as hmac_mod
import json
import logging
import time
from typing import Optional

import httpx

from hiitl.core.types import (
    CostEstimate,
    Decision,
    DecisionType,
    Sensitivity,
    Timing,
)
from hiitl.sdk.config import HostedModeConfig
from hiitl.sdk.exceptions import NetworkError, ServerError

logger = logging.getLogger(__name__)

# HTTP status codes that are retryable (transient failures)
_RETRYABLE_STATUS_CODES = {502, 503, 504, 429}


class HostedClient:
    """HTTP client for ECP server hosted mode.

    Manages HTTP communication with the ECP server, including
    authentication, retries, and response parsing.
    """

    def __init__(self, config: HostedModeConfig):
        self._config = config
        self._client = httpx.Client(
            base_url=config.server_url,
            timeout=config.timeout,
            headers={
                "Authorization": f"Bearer {config.api_key}",
                "Content-Type": "application/json",
            },
        )

    def evaluate(
        self,
        action: str,
        operation: str,
        target: dict,
        parameters: dict,
        *,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        sensitivity: Optional[list[Sensitivity]] = None,
        cost_estimate: Optional[CostEstimate] = None,
        reason: Optional[str] = None,
        envelope_signature: Optional[str] = None,
    ) -> Decision:
        """Send evaluate request to ECP server and return Decision.

        Builds the request body matching the server's EvaluateRequest schema,
        sends it with retry logic, and parses the response into a Decision.

        Args:
            action: Action name
            operation: Operation type
            target: Target resource
            parameters: Operation parameters
            user_id: User identifier (optional)
            session_id: Session identifier (optional)
            sensitivity: Sensitivity labels (optional)
            cost_estimate: Cost estimate (optional)
            reason: Reasoning for action (optional)
            envelope_signature: HMAC-SHA256 signature (optional)

        Returns:
            Decision object from server response

        Raises:
            ServerError: Server returned an error response
            NetworkError: Cannot reach the server
        """
        # Build request body (EvaluateRequest format)
        body: dict = {
            "action": action,
            "operation": operation,
            "target": target,
            "parameters": parameters,
        }

        if self._config.agent_id:
            body["agent_id"] = self._config.agent_id
        if user_id:
            body["user_id"] = user_id
        if session_id:
            body["session_id"] = session_id
        if sensitivity:
            body["sensitivity"] = [
                s.value if isinstance(s, Sensitivity) else s
                for s in sensitivity
            ]
        if cost_estimate:
            body["cost_estimate"] = (
                cost_estimate.model_dump()
                if hasattr(cost_estimate, "model_dump")
                else cost_estimate
            )
        if reason:
            body["reason"] = reason

        # Compute envelope signature if signing key is configured
        if envelope_signature:
            body["envelope_signature"] = envelope_signature
        elif self._config.signature_key:
            body["envelope_signature"] = self._compute_signature(body)

        # Send with retry
        response = self._send_with_retry(body)

        # Parse response to Decision
        return self._parse_response(response)

    def _compute_signature(self, body: dict) -> str:
        """Compute HMAC-SHA256 signature of request body."""
        canonical = json.dumps(body, sort_keys=True, separators=(",", ":"))
        return hmac_mod.new(
            self._config.signature_key.encode(),
            canonical.encode(),
            hashlib.sha256,
        ).hexdigest()

    def _send_with_retry(self, body: dict) -> httpx.Response:
        """Send POST request with exponential backoff retry on transient failures."""
        last_error: Exception | None = None

        for attempt in range(self._config.max_retries + 1):
            try:
                response = self._client.post("/v1/evaluate", json=body)

                # Success or non-retryable error: return immediately
                if response.status_code < 500 and response.status_code != 429:
                    return response

                # Retryable server error
                if response.status_code in _RETRYABLE_STATUS_CODES:
                    if attempt < self._config.max_retries:
                        wait = _backoff_delay(attempt)
                        logger.warning(
                            "ECP server returned %d, retrying in %.1fs (attempt %d/%d)",
                            response.status_code,
                            wait,
                            attempt + 1,
                            self._config.max_retries,
                        )
                        time.sleep(wait)
                        continue

                # Non-retryable 5xx or exhausted retries
                return response

            except httpx.TimeoutException as e:
                last_error = e
                if attempt < self._config.max_retries:
                    wait = _backoff_delay(attempt)
                    logger.warning(
                        "ECP server request timed out, retrying in %.1fs (attempt %d/%d)",
                        wait,
                        attempt + 1,
                        self._config.max_retries,
                    )
                    time.sleep(wait)
                    continue

            except httpx.ConnectError as e:
                raise NetworkError(self._config.server_url, e) from e

            except httpx.HTTPError as e:
                raise NetworkError(self._config.server_url, e) from e

        # All retries exhausted on timeout
        raise NetworkError(self._config.server_url, last_error)

    def _parse_response(self, response: httpx.Response) -> Decision:
        """Parse HTTP response into a Decision or raise appropriate error."""
        if response.status_code == 200:
            data = response.json()

            # Map server timing format to SDK Timing model
            # Server returns {"total_ms": X, ...} — SDK expects ingest_ms + evaluation_ms
            raw_timing = data.get("timing", {})
            total_ms = raw_timing.get("total_ms", 0.0)
            timing = Timing(
                ingest_ms=raw_timing.get("ingest_ms", 0.0),
                evaluation_ms=raw_timing.get("evaluation_ms", total_ms),
                total_ms=total_ms,
            )

            # Pass all optional fields through — Pydantic validates and
            # coerces types (e.g. error dict → ErrorDetail via model_validator).
            # Use .get() with None default; Pydantic ignores None optionals.
            _OPT_FIELDS = (
                "envelope_hash", "resume_token", "route_ref",
                "escalation_context", "matched_rules", "rate_limit",
                "approval_metadata", "sandbox_metadata", "error",
                "remediation", "would_be", "would_be_reason_codes",
            )
            kwargs = {
                k: data[k] for k in _OPT_FIELDS
                if k in data and data[k] is not None
            }

            return Decision(
                action_id=data.get("action_id", ""),
                decision=DecisionType(data["decision"]),
                allowed=data["allowed"],
                reason_codes=data.get("reason_codes", []),
                policy_version=data.get("policy_version", ""),
                timing=timing,
                **kwargs,
            )

        # Error response — extract structured error info
        error_code = "UNKNOWN_ERROR"
        message = f"Server returned HTTP {response.status_code}"

        try:
            data = response.json()
            # Server wraps errors in {"detail": {...}}
            detail = data.get("detail", data)
            if isinstance(detail, dict):
                error_code = detail.get("error", error_code)
                message = detail.get("message", message)
            elif isinstance(detail, str):
                message = detail
        except Exception:
            message = response.text or message

        raise ServerError(response.status_code, error_code, message)

    def close(self):
        """Close the HTTP client and release connections."""
        self._client.close()


def _backoff_delay(attempt: int) -> float:
    """Exponential backoff: 0.5s, 1s, 2s, capped at 4s."""
    return min(0.5 * (2 ** attempt), 4.0)
